### ISSUE TITLE
Android DNS does not resolve

### ISSUE THREAD

**Issue by tdalbo92** [tailscale/tailscale#2102]
As of an update last week, DNS queries do not succeed on my Pixel.  My DNS settings in Tailscale are as such:

- Magic DNS is not enabled
- nameserver set to my Pi running pihole's Tailscale IP
- Split domain set to ^^ IP

I'm had been using pihole to resolve local domains to their assigned Tailscale IPs, which allowed me to access devices at home from my phone when away from the house via FQDN.  Without Tailscale on Android working, I just leave it off and lose access to things when I'm away from home.

**Comment by DentonGentry** [comment]
With https://login.tailscale.com/admin/dns of:
Magic DNS disabled
Split DNS: ts-dns.test 100.120.55.47
Global DNS: 100.92.61.18 with Override local enabled

Android is unable to look up constantine.ts-dns.test from the Split DNS server:

```
% dscacheutil -q host -a name constantine.ts-dns.test
%
```

I've been able to look up other, global DNS names using 100.100.100.100, like yahoo.com and google.com.

**Comment by DentonGentry** [comment]
@tdalbo92 could you send a screenshot of https://login.tailscale.com/admin/dns to support@tailscale.com, and reference this bug number? I haven't been able to produce a similar failure with global servers, I can only produce a failure with the split domain.

**Comment by DentonGentry** [comment]
Brainstorming from Eng meeting: sounds like Fallback DNS isn't being pushed to the Android client.

**Comment by normanr** [comment]
also see last few entries from #1956, I suspected that upstream DNS requests (sent by the Android client) are not being routed onto the tailscale network.

**Comment by DentonGentry** [comment]
With https://login.tailscale.com/admin/dns of:
Magic DNS disabled
Split DNS: ts-dns.test 100.120.55.47
Global DNS: 100.92.61.18 with Override local enabled

Android is unable to look up constantine.ts-dns.test from the Split DNS server, using Chrome.

I've been able to look up other, global DNS names using 100.100.100.100, like yahoo.com and google.com.

adb logcat does show the DNS settings appearing, which looks correct:
06-15 21:28:11.853  2858  2888 I com.tailscale.ipn: 10.7M/101.9M dns: Resolvercfg: {Routes:map[.:[100.92.61.18:53] ts-dns.test.:[100.120.55.47:53]] Hosts:map[...

**Comment by danderson** [comment]
got a tailscale bugreport I can look at for logs?

**Comment by DentonGentry** [comment]
We don't have a way to generate a bugreport from Android, but at 6:29am Pacific time 6/16/2021 I rebooted the device, started a Tailscale app I built from HEAD this morning, and tried to reach constantine.ts-dns.test using Chrome and DNSdig.
Logid 3d02c361cc296fe328a120e3075565613ae39bb8bbeafb4573413eaa983eb6b0

With https://login.tailscale.com/admin/dns of:
Magic DNS enabled
Split DNS: ts-dns.test 100.120.55.47
Global DNS: 100.92.61.18 with Override local enabled

**Comment by DentonGentry** [comment]
Adding some logging, I can see queries for the split domain being sent to the correct IP address for that resolver.

```
06-16 21:20:41.966 14820 14850 I com.tailscale.ipn: 8.9M/104.7M dns: dns: forward: forward nameFromQuery: connectivitycheck.gstatic.com.
06-16 21:20:41.966 14820 14850 I com.tailscale.ipn: 8.9M/104.7M dns: dns: forward: send: connectivitycheck.gstatic.com. to 100.92.61.18:53
06-16 21:20:47.930 14820 14885 I com.tailscale.ipn: 9.1M/105.0M dns: dns: forward: forward nameFromQuery: constantine.ts-dns.test.
06-16 21:20:47.931 14820 14885 I com.tailscale.ipn: 9.1M/105.0M dns: dns: forward: send: constantine.ts-dns.test. to 100.120.55.47:53
06-16 21:20:48.522 14820 14885 I com.tailscale.ipn: 9.1M/105.0M dns: dns: forward: forward nameFromQuery: app-measurement.com.
06-16 21:20:48.522 14820 14885 I com.tailscale.ipn: 9.1M/105.0M dns: dns: forward: send: app-measurement.com. to 100.92.61.18:53
06-16 21:20:48.524 14820 14890 I com.tailscale.ipn: 9.1M/105.0M dns: dns: forward: forward nameFromQuery: app-measurement.com.
06-16 21:20:48.524 14820 14890 I com.tailscale.ipn: 9.1M/105.0M dns: dns: forward: send: app-measurement.com. to 100.92.61.18:53
```

However running tcpdump on the tailscale interface of 100.120.55.47, there are no DNS queries from the Android device.

**Comment by DentonGentry** [comment]
Adding even more debugging, Android thinks the send was successful:

```
06-16 21:38:25.196 16123 16158 I com.tailscale.ipn: 9.1M/106.3M dns: dns: forward: forward nameFromQuery: constantine.ts-dns.test.
06-16 21:38:25.196 16123 16158 I com.tailscale.ipn: 9.1M/106.3M dns: dns: forward: send: constantine.ts-dns.test. to 100.120.55.47:53
06-16 21:38:25.197 16123 16158 I com.tailscale.ipn: 9.1M/106.3M dns: dns: forward: fwdConn 21: send: success
```

**Comment by DentonGentry** [comment]
Tried reverting https://github.com/tailscale/tailscale/commit/6def64751446bba6914cccd51f5f4640298a3a5f just to check, but it made no difference.

The same code path when sending to the global resolver 100.92.61.18 is working, but when it sends to 100.120.55.47 the DNS packet never arrives at 100.120.55.47.

**Comment by DentonGentry** [comment]
If I set up 100.120.55.47 as a DNS server in the Android app DNS dig and query constantine.ts-dns.test, it works.

If I query 100.100.100.100, the DNS packet never arrives at 100.120.55.47.

**Comment by DentonGentry** [comment]
Packet is going out the wrong interface.

The Split DNS server is at

**Comment by DentonGentry** [comment]
My global DNS server is set to 100.92.61.18. Those packets are also going out the wrong interface, to Wi-Fi instead of the Tailscale VPN. The difference is that 100.92.61.18 is the Tailscale address of my router, and even though the packet is coming in the wrong interface the router still recognizes one of its IP addresses and responds.

So that one is broken too, just not in an immediately obvious way.

**Comment by DentonGentry** [comment]
Running a shell from adb, the Android kernel's routing table shows only a default route. I think this is normal for Android, the VPNBuilder doesn't add routes to the kernel. I just don't really understand how the VPNBuilder does get packets to go over the VPN, and what we'd need to do to allow some, though not all, of the packets the Tailscale APK itself sends to go over the VPN.

```
bonito:/ $ netstat -rn
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
10.1.10.0       0.0.0.0         255.255.255.0   U         0 0          0 wlan0
bonito:/ $
```

**Comment by DentonGentry** [comment]
This happens because we tell the VpnBuilder to exclude our own app ID: https://github.com/tailscale/tailscale-android/blob/9e9d69fd95bc0080a111aefcdd3b5a9e73d89056/android/src/main/java/com/tailscale/ipn/IPNService.java#L65
As our packets are not allowed to go into the tunnel... they don't go into the tunnel and go out the Wi-Fi instead.

Removing the call to addDisallowedApplication() makes my test case work, I can resolve constantine.ts-dns.test and I no longer see the 100.x.x.x packets on the Wi-Fi network. I suspect the reason we were calling addDisallowedApplication is to ensure that our connections to the coordination server and log server don't try to go over the VPN. Trying to use Tailscale to connect to Tailscale wouldn't work.

We'll evaluate negative side-effects from removing the addDisallowedApplication() call. The main concern will be whether it will impact exit nodes on Android, where we send all traffic via the VPN. Without the call to addDisallowedApplication(), even the coordination server connections will use the VPN. On other platforms we use network namespaces https://pkg.go.dev/tailscale.com@v1.1.0/net/netns, but I need to check whether this is sufficient on Android.

**Comment by DentonGentry** [comment]
I ran a packet capture on the Wi-Fi router while running an Android app with the addDisallowedApplication() call removed.

When I turned on an exit node in Android, I stopped seeing responses from the Android's IP address back to control at 3.121.18.47:443. I see packets coming from control trying to reach the Android address, presumably retries from before the exit node was enabled, but no responses or even ACKs back from Android.

I think we have to call `addDisallowedApplication()` to allow exit nodes to work.

We may be able to make Tailscale IP addresses in DNS configs work by doing our own routing lookup and sending packets directly to wireguard as appropriate, bypassing Android's notion of what route we should use.

**Comment by DentonGentry** [comment]
I think this will end up being the same fix as https://github.com/tailscale/tailscale/issues/2128 and https://github.com/tailscale/tailscale/issues/1887:

We'll need to make MagicDNS consult an internal routing table to decide whether the destination is public and it should use a regular socket, or whether the destination is reached via Tailscale and we need to hand the packet straight to wireguard-go. Note that the check isn't just going to be for 100.64/10, #1887 concerns cases using subnet relays where non-Tailscale addresses are supposed to traverse the Tailnet.

Getting the packet sent via wireguard-go is the easier part, grabbing the responses out of wireguard-go to hand back to Magic DNS is likely to be more difficult.

**Comment by normanr** [comment]
btw, for Android try `ip route ls table all` (and `ip rule ls`, to see what gets routed to which tables).

ahh, so it's _exactly_ what I suspected. In order to receive packets, what about sending the dns requests directly to wireguard-go, but use one of the fwdConn conn's address/port as the source address. That way the replies might just get routed by the kernel back to the tailscale socket? Android doesn't seem to enable Reverse Path Filtering, so this has a high chance of working, worse case you could have a set of dummy sockets, and you just intercept replies coming from wireguard-go because sending the on to the vpn interface.

**Comment by eliasnaur** [comment]
> This happens because we tell the VpnBuilder to exclude our own app ID: https://github.com/tailscale/tailscale-android/blob/9e9d69fd95bc0080a111aefcdd3b5a9e73d89056/android/src/main/java/com/tailscale/ipn/IPNService.java#L65
> As our packets are not allowed to go into the tunnel... they don't go into the tunnel and go out the Wi-Fi instead.
> 
> Removing the call to addDisallowedApplication() makes my test case work, I can resolve constantine.ts-dns.test and I no longer see the 100.x.x.x packets on the Wi-Fi network.
> 
> ```
> $ git diff android/src/main/java/com/tailscale/ipn/IPNService.java
> diff --git a/android/src/main/java/com/tailscale/ipn/IPNService.java b/android/src/main/java/com/tailscale/ipn/IPNService.java
> index 875adf5..7e00814 100644
> --- a/android/src/main/java/com/tailscale/ipn/IPNService.java
> +++ b/android/src/main/java/com/tailscale/ipn/IPNService.java
> @@ -62,12 +62,6 @@ public class IPNService extends VpnService {
>                         .setConfigureIntent(configIntent())
>                         .allowFamily(OsConstants.AF_INET)
>                         .allowFamily(OsConstants.AF_INET6);
> -               try {
> -                       b.addDisallowedApplication(BuildConfig.APPLICATION_ID);
> -               } catch (PackageManager.NameNotFoundException e) {
> -                       // This error means com.tailscale.ipn isn't
> -                       // installed. That shouldn't happen, so pretend it didn't.
> -               }
>                 if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
>                         b.setMetered(false); // Inherit the metered status from the underlying networks.
>                 if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
> ```
> 
> We call addDisallowedApplication to ensure that our connections to the coordination server and log server don't try to go over the VPN. We'll evaluate negative side-effects from removing the addDisallowedApplication() call. The main concern will be whether it will impact exit nodes on Android, where we send all traffic via the VPN. Without the call to addDisallowedApplication(), even the coordination server connections would use the VPN. On other platforms we use network namespaces https://pkg.go.dev/tailscale.com@v1.1.0/net/netns, but I need to check whether this is sufficient on Android.

Good find! I completely missed that.

Can we use VPNService.protect on the sockets that need to avoid the tunnel?

**Comment by bradfitz** [comment]
> Can we use VPNService.protect on the sockets that need to avoid the tunnel?

That's effectively what we do on Linux and Windows with the `net/netns` package. One route we're exploring is looking into how we can do the same on Android. @danderson might remember why we weren't able to do that in the past.

**Comment by normanr** [comment]
fwiw, Android's APIs to bind a process/socket to a different interface are: [ConnectivityManager.bindProcessToNetwork()](https://developer.android.com/reference/android/net/ConnectivityManager#bindProcessToNetwork(android.net.Network)) and [Network.bindSocket()](https://developer.android.com/reference/android/net/Network#bindSocket(java.net.Socket)), but the VPN must allow bypass first (I assume [VpnService.protect()](https://developer.android.com/reference/android/net/VpnService#protect(int)) doesn't require this, but it can only be used by the application providing the VPN service). ref: https://developer.android.com/guide/topics/connectivity/vpn#bypass_vpn

**Comment by bradfitz** [comment]
@eliasnaur, I've fixed this issue and temporarily pushed the results to https://github.com/tailscale/tailscale-android/commits/bradfitz/1.9

There are three commits of interest:

* https://github.com/tailscale/tailscale/commit/7b8ed1fc094e769f88dd37bf7fd9188af4d74bf9 adds the hook in the `tailscale.com/net/nets` package to mark file descriptors as not using the VPN. It's only in main and not (yet?) backported to 1.8. It's small enough to backport but also probably not worth it, considering we want to get 1.10 out ~today.
* https://github.com/tailscale/tailscale-android/commit/97f973cfc2fbb69d289de29cfb15120cad4ad4c2 updates Android from 1.8.6 to main (1.9.x). You probably have this code already privately somewhere if you've been working on file stuff.
* https://github.com/tailscale/tailscale-android/commit/2bad4b2b2e706dfed4d7ed11b13d8e2e400ecd30 is the actual fix. It stops using `addDisallowedApplication` (from https://github.com/tailscale/tailscale-android/commit/8daee9c431264c7780a39802e6213a1b13cab7f8) and instead registers the new hook in the `net/netns` package. You may want to double-check my JNI; it's never been my forte.

How do you want to proceed with branches for Android? I think we should probably have a release-branch/1.8 (and release-branch/1.10 as well) for Android and then unstable development can continue on main, matching the main repo.

**Comment by bradfitz** [comment]
@normanr, thanks for all the pointers, btw!

**Comment by eliasnaur** [comment]
Thank you for finding and fixing this issue. I've released a new version with the fixes and the 1.10 upgrade to Play Store.

> How do you want to proceed with branches for Android? I think we should probably have a release-branch/1.8 (and release-branch/1.10 as well) for Android and then unstable development can continue on main, matching the main repo.

Release branches wouldn't help me, and I prefer to keep main stable and have feature branches (FWIW, the public `file-sharing` branch included my version of the tailscale 1.9 upgrade). I suppose whoever takes over app development can introduce a new branching scheme and I'll follow that.
